syntax = "proto3";

package halo.pb;

///////////////////// Data definitions ///////////////////////

// Ideally this would also contain information about blocks in the function.
message FunctionInfo {
  string label = 1;
  uint32 size = 2;
  uint64 start = 3;
}

message ProcessInfo {
  bytes bitcode = 1;
}



message BranchInfo {
  uint64 from = 1;
  uint64 to = 2;
  bool mispred = 3;
  bool predicted = 4;
}

////////////////////////////////////////////////////////////////



//////////////////////  Message Definitions  ///////////////////


///////////////////// Client -> Server /////////////////////////


// First message expected by the server upon connection with information about
// the client.
message ClientEnroll {
  string exe_path = 1;
  string process_triple = 2;
  string host_cpu = 3;
  string build_cmd = 4;
  // TODO: some sort of SHA hash of the object file. This way clients running
  // identical code don't have to send bitcode if the server already has it.

  repeated FunctionInfo funcs = 5;
}


message RawSample {
  uint64 instr_ptr = 1;
  uint32 thread_id = 2;
  uint64 time = 3;
  repeated uint64 call_context = 4;
  repeated BranchInfo branch = 5;
}
